plutus-core/plutus-core/src/PlutusCore/ParserCommon.hs

```diff
@@ -0,0 +1,179 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# LANGUAGE TypeApplications  #-}
+{-# LANGUAGE TypeOperators     #-}
+
+-- | Common functions for parsers of UPLC, PLC, and PIR.
+module PlutusCore.ParserCommon where
+
+import           Data.Char                  (isAlphaNum)
+import qualified Data.Map                   as M
+import qualified Data.Text                  as T
+import           ErrorCode                  (ErrorCode (ErrorCode), HasErrorCode (..))
+import qualified PlutusCore                 as PLC
+import qualified PlutusCore.Parsable        as PLC
+import           PlutusPrelude
+import           Text.Megaparsec            hiding (ParseError, State, parse)
+import qualified Text.Megaparsec            as Parsec
+import           Text.Megaparsec.Char       (char, letterChar, space1, string)
+import qualified Text.Megaparsec.Char.Lexer as Lex
+
+import           Control.Monad.State        (MonadState (get, put), StateT, evalStateT)
+
+import           Data.Proxy                 (Proxy (Proxy))
+
+data ParseError = UnknownBuiltinType T.Text
+                | InvalidConstant T.Text T.Text
+                deriving (Eq, Ord, Show)
+
+instance HasErrorCode ParseError where
+      errorCode UnknownBuiltinType {} = ErrorCode 5
+      errorCode InvalidConstant {}    = ErrorCode 4
+
+type Error = Parsec.ParseError Char ParseError
+
+instance ShowErrorComponent ParseError where
+    showErrorComponent (UnknownBuiltinType ty) = "Unknown built-in type: " ++ T.unpack ty
+    showErrorComponent (InvalidConstant ty con) =
+        "Invalid constant: " ++ T.unpack con ++ " of type " ++ T.unpack ty
+newtype ParserState = ParserState { identifiers :: M.Map T.Text PLC.Unique }
+    deriving (Show)
+
+initial :: ParserState
+initial = ParserState M.empty
+
+intern :: (MonadState ParserState m, PLC.MonadQuote m) => T.Text -> m PLC.Unique
```

Comment:
Let's try and get haddock on all most of the non-trivial functions.

Reply:
I've done a couple, lmk if you think more is needed.